cli/main.c

```diff
@@ -23,157 +24,51 @@
-        error(EXIT_FAILURE, errno, "unable to read the peer key from %s",
-              argv[2]);
-      }
-      print_hex(stderr, "peer-key:   0x", peer_key, sizeof peer_key);
-    case 2:
-      if (read_key(argv[1], private_key, sizeof private_key) !=
-          sizeof private_key) {
-        if (errno != ENOENT) {
-          error(EXIT_FAILURE, errno, "unable to read the private key from %s",
-                argv[1]);
-        }
-        int fd = open(argv[1], O_WRONLY | O_EXCL | O_CREAT, S_IRUSR | S_IWUSR);
-        if (fd == -1) {
-          error(EXIT_FAILURE, errno, "unable to create a new file at %s",
-                argv[1]);
-        }
-        if (glome_generate_key(private_key, public_key)) {
-          error(EXIT_FAILURE, 0, "unable to generate a new key pair");
-        }
-        if (write(fd, private_key, sizeof private_key) != sizeof private_key) {
-          error(EXIT_FAILURE, errno, "unable to write the private key to %s",
-                argv[1]);
-        }
-        close(fd);
-      }
-      break;
-    default:
-      error(EXIT_FAILURE, 0, kUsage, argv[0], argv[0], argv[0]);
-  }
+typedef int (*mainfunc)(int, char **);
 
-  if (glome_derive_key(private_key, public_key)) {
-    error(EXIT_FAILURE, 0, "unable to derive public key");
-  }
+struct cmd {
+  const char *name;
+  mainfunc run;
+};
 
-  if (argc == 2) {  // key generation
-    if (isatty(STDOUT_FILENO) == 1) {
-      print_hex(stdout, NULL, public_key, sizeof public_key);
-    } else {
-      write(STDOUT_FILENO, public_key, sizeof public_key);
-    }
-    return EXIT_SUCCESS;
-  }
+static const struct cmd cmds[] = {
+    {"genkey", &genkey}, {"pubkey", &pubkey}, {"tag", &tag},
+    {"verify", &verify}, {NULL, &print_help},
+};
 
-  // Debug information
-  print_hex(stderr, "public-key: 0x", public_key, sizeof public_key);
-  fprintf(stderr, "message:   '%s'\n", message == NULL ? "" : message);
-  fprintf(stderr, "counter:    %ld\n", counter);
-  fprintf(stderr, "verify:     %d\n", verification);
-
-  if (glome_tag(verification, counter, private_key, peer_key,
-                (uint8_t *)message, message == NULL ? 0 : strlen(message),
-                tag)) {
-    error(EXIT_FAILURE, 0, "MAC tag generation failed");
+int main(int argc, char **argv) {
+  if (argc == 0) {
+    fputs("called with empty argv\n", stderr);
+    return EXIT_FAILURE;
   }
-
-  if (argc == 6) {  // tag verification
-    print_hex(stderr, "mac-tag:    0x", tag, sizeof tag);
-    print_hex(stderr, "unverified: 0x", expected_tag, expected_tag_len);
-    if (memcmp(expected_tag, tag, expected_tag_len) != 0) {
-      error(EXIT_FAILURE, 0, "MAC tag verfication failed");
-    }
-  } else {  // tag generation
-    print_hex(stdout, NULL, tag, sizeof tag);
+  if (argc < 2) {
+    return print_help(argc, argv);
   }
-
-  return EXIT_SUCCESS;
+  // Advance to the requested subcommand.
+  const struct cmd *c;
+  for (c = cmds; c->name && strcmp(c->name, argv[1]); c++)
+    ;
```

Comment:
That's almost "smart" code. Could you add a comment to `cmds[]` that the last one will be executed when none is found? Also I think you always need to start a block, although I'm somewhat unsure about this. I almost feel like a while loop might be easier to read.

Reply:
Is that a praise or an insult? :D Added comments, and it is much more readable with a while loop.